### PR TITLE
[Merged by Bors] - perf: use implicit parameters in `SMul` hierarchy

### PR DESCRIPTION
**Pull request by astrainfinita**
---

This is the part of #16646 (or #16486) that is not resolved by https://github.com/leanprover/lean4/pull/5376.

split from #17411

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by Vierkantor — APPROVED** [review]
So this review is going to mirror my comment here closely: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Ways.20to.20speed.20up.20Mathlib/near/471549289

First of all, a negative comment: I don't like that we have to specify implicitness manually, since it hurts the readability somewhat. So let's see if the advantages justify the change.

This change is very similar to what we had for some time in Mathlib 3 for modules, and has essentially the same motivation. This is still somewhat of a workaround for the fact that controlling synthesis order precisely is hard, but at least we know from experience that it can work (and when it doesn't). In particular, we used to run into problems when we were working with modules over rings. Then `[CommRing R] [AddCommGroup M]` can't always be inferred from the `[Module R M]` instance, so they would have to switch between `{implicit}` and `[instance]` parameters many times in the same file. Since the current modifications only touch projections from one structure to another, that should not be any issue.

Moreover, if Lean becomes smarter about instance synth order, we should be able to undo the change just like we did for Mathlib 3. So there is not too much harm in going for the change.

One thing that I'm a bit worried about is that we can't run a benchmark, and this is a `perf` PR. Since we do have a benchmark on #17411 and that is very green, I am willing to accept this based on the experience that says it should work, and the assumption that doing only a part is not going to cause any regressions.

In summary: I think this is an okay workaround where the results justify a mild reduction in readability. Let's Get This Merged!

bors r+

**Comment by grunweg** [comment]
!bench
